### PR TITLE
[Github] Fix Rate limit error

### PR DESCRIPTION
**Pull request by moxarth-elastic**
### Relates to https://github.com/elastic/connectors/issues/2636

This PR fixes the rate limiting issue in the Github connector while facing a 403 status code (Forbidden). The connector is able to fetch the documents after token has hit the rate limit.

#### Pre-Review Checklist
- [x] this PR does NOT contain credentials of any kind, such as API keys or username/passwords (double check `config.yml.example`)
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference
- [ ] if you added or changed Rich Configurable Fields for a Native Connector, you made a corresponding PR in [Kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/native_connectors.ts)

## Release Note

Fixing the rate limit issue while facing 403 - Forbidden error.

### CODE REVIEW

File: connectors/sources/github.py [hunk 1]
```diff
@@ -801,11 +799,20 @@ async def graphql(self, query, variables=None):
             msg = "Your Github token is either expired or revoked. Please check again."
             raise UnauthorizedException(msg) from exception
         except BadGraphQLRequest as exception:
-            if self.get_rate_limit_encountered(exception.status_code, exception):
-                await self._put_to_sleep(resource_type="graphql")
```
Comment (navarone-feekery) [inline connectors/sources/github.py:805]: Why do we no longer want to check rate limiting after a `BadGraphQLRequest` exception?

File: connectors/sources/github.py [hunk 2]
```diff
@@ -873,19 +884,39 @@ async def paginated_api_call(self, query, variables, keys):
...
+                "HEAD", url, request_headers
+            )
+            scopes = headers.get("X-OAuth-Scopes")
+            if not scopes or not scopes.strip():
+                self._logger.warning(
+                    f"Couldn't find 'X-OAuth-Scopes' in headers {headers}"
+                )
+                return set()
+            return {scope.strip() for scope in scopes.split(",")}
+        except Exception as exception:
```
Comment (navarone-feekery) [inline connectors/sources/github.py:911]: Is there not a specific exception to capture here? I think it's worthwhile to separate specific error exceptions and unexpected exceptions

File: connectors/sources/github.py [hunk 3]
```diff
@@ -1291,17 +1322,23 @@ async def _get_invalid_repos_for_github_app(self):
...
-                continue
-            owner, repo_name = self.github_client.get_repo_details(
-                repo_name=full_repo_name
-            )
-            if await self._get_repo_object_for_github_app(owner, repo_name) is None:
-                invalid_repos.add(full_repo_name)
-
-        return list(invalid_repos)
+            return list(invalid_repos)
+        except Exception as exception:
```
Comment (navarone-feekery) [inline connectors/sources/github.py:1336]: Same as above; is there a specific exception we can try to catch first?

**Review by navarone-feekery — COMMENTED** [review]
Sorry for the delay in reviewing.
I want to confirm, does this PR actually close the issue https://github.com/elastic/connectors/issues/2636 ? Because reading that issue, it sounds like a somewhat different problem.

**Comment by moxarth-elastic** [comment]
> I want to confirm, does this PR actually close the issue https://github.com/elastic/connectors/issues/2636 ? Because reading that issue, it sounds like a somewhat different problem.

I have updated the description; added https://github.com/elastic/connectors/issues/2636 for `Relates to` tag. Because i think the issue has happened due to Rate limiting issue; not because of a private repository. We even informed this issue to the reporter - https://github.com/elastic/connectors/issues/2636#issuecomment-2210774762.

File: connectors/sources/github.py (hunk 1 above)
Comment (moxarth-elastic) [reply connectors/sources/github.py:805]: This function raises `QueryError` instead of `BadGraphQLRequest` when rate limit is encountered.

**Review by navarone-feekery — APPROVED** [review]
Nice work 🚀

**Comment by seanstory** [comment]
Since it's a bugfix, adding a label for the most recent minor to ensure a backport.

**Comment by seanstory** [comment]
@moxarth-elastic can you run the ☝️ above command, resolve the conflicts, and be sure to open the 8.15 backport?

**Comment by moxarth-elastic** [comment]
> @moxarth-elastic can you run the ☝️ above command, resolve the conflicts, and be sure to open the 8.15 backport?

Backport PR Link - https://github.com/elastic/connectors/pull/2746

**Pull request merged**